### ISSUE TITLE
NamedTuple _replace calls are treated as private accessors

### ISSUE THREAD

**Issue by refi64** [microsoft/pyright#1589]
**Describe the bug**

All namedtuple instances have a method `_replace` that is intended to be used to replace fields. This is reported by pyright as an error.

**To Reproduce**

Simple example:

```python
from typing import NamedTuple

class A(NamedTuple):
    a: int

A(a=1)._replace(a=2)
```

**Expected behavior**

This produces an error:

```
"_replace" is protected and used outside of the class in which it is declared (reportPrivateUsage)
```

This also happens with `urlparse`'s result:

```
urllib.parse.urlparse('abc')._replace(fragment='')
```

**VS Code extension or command-line**

Command-line version is 1.1.118, VSCode version is 1.1.117 included with PyLance.

**Comment by erictraut** [comment]
This is intended. Methods named with an underscore are assumed to be private.

If you want to use the `_replace` method, you will need to disable the `reportPrivateUsage` diagnostic check for that file or line.

**Comment by refi64** [comment]
Is there any reason this couldn't somehow be whitelisted? I understand it starts with an underscore and looks private, but this function is on pretty much every namedtuple, so having to put ignored diagnostics on every `_replace` usage to avoid just flipping it off everywhere wouldn't be very fun :sweat_smile:

**Comment by erictraut** [comment]
In general, it's not safe to call `_replace` on a class that derives from namedtuple, which is why it is marked as "protected". If you want to use it and you know that it's safe in your particular case, you can suppress the error.

For more details, see [this post](https://softwareengineering.stackexchange.com/questions/315348/why-is-the-replace-method-of-python-namedtuple-classes-protected#:~:text=It%20is%20not%20a%20base%20class%2C%20it%20is%20a%20class%20generator.&text=So%20the%20namedtuple%20might%20be,classes%20might%20want%20that%20functionality.).

Pyright generally does not have internal knowledge of specific types. It bases all of its checks on type information provided to it in type stubs. It would be a slippery slope to start adding special-case knowledge of the internal semantics of specific classes.
